Advanced Building Blocks - Enumerable Methods[SOLO]

tirthajyoti-ghosh opened the pull request:
- Built-in enum methods such as: `#each`, `#each_with_index`, `#select`, `#all?`, `#any?`, `#none?`, `#count`, `#map`, `#inject`  re-written in custom algorithm.
- Made separate file for testing - `enum_test.rb`. `enumerables.rb` file containing actual method codes.
- Method names are like this: `my_` followed built-in method names.
- `#multiply_els` created in `enum_test.rb` file to test `#my_inject` as per project requirements.

tirthajyoti-ghosh committed:
added #my_each

tirthajyoti-ghosh committed:
added #my_each_with_index

tirthajyoti-ghosh committed:
added #my_select

tirthajyoti-ghosh committed:
added #my_all?, #my_any?, #my_none

tirthajyoti-ghosh committed:
added #my_count

tirthajyoti-ghosh committed:
added #my_inject and #my_map

tirthajyoti-ghosh committed:
fixed #my_any?. added test file

tirthajyoti-ghosh committed:
fixed #my_none?, #my_count, #my_map. added test for remaining methods.

tirthajyoti-ghosh committed:
updated README

tirthajyoti-ghosh committed:
updated README - fixed issues link

tirthajyoti-ghosh committed:
updated README - removed issues link

tirthajyoti-ghosh committed:
fixed enum_test.rb

whiz25 reviewed (changes requested):
## Nice Progress :+1: 
You have made amazing progress with the project. Your code covers most edge cases but we need to fix a few failing ones cases before we can merge.
Check my comments for suggestions.

### Martin Afani

whiz25 commented on enumerables.rb:
```diff
@@ -0,0 +1,114 @@
+      i += 1
+    end
+    true
+  end
```
- [ ] should be identical and returns the same thing as ruby's [`all?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_all.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [ ] when a class is passed as an argument returns true if all of the collection is a member of such class
```ruby
array = Array.new(100) { rand(0...9) }
array.my_all?(Integer) == array.all?(Integer)
```
- [ ] when a Regex is passed as an argument returns true if all of the collection matches the Regex
```ruby
words = %w[dog door rod blade]
words.my_all?(/d/) == words.all?(/d/)
```
- [ ] when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern
```ruby
array.my_all?(3) == array.all?(3) #true

whiz25 commented on enumerables.rb:
```diff
@@ -0,0 +1,114 @@
+      i += 1
+    end
+    false
+  end
```
- [ ] should be identical and returns the same thing as ruby's [`any?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-any-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_any.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [ ] when no block or argument is given returns true if at least one of the collection is not false or nil
```ruby
true_array = [nil, false, true, []] 
false_array = [nil, false, nil, false]
true_array.my_any? == true_array.any? #true
```
- [ ]  when a class is passed as an argument returns true if at least one of the collection is a member of such class
```ruby
words = %w[dog door rod blade]
words.my_any?(Integer) == words.any?(Integer) #true
```
- [ ] when a Regex is passed as an argument returns false if none of the collection matches the Regex
```ruby
words.my_any?(/z/) == words.any?(/z/) #true
```
- [ ] when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern
```ruby
words.my_any?('cat') == words.any?('cat') #true

whiz25 commented on enumerables.rb:
```diff
@@ -0,0 +1,114 @@
+      i += 1
+    end
+    true
+  end
```
- [ ] should be identical and returns the same thing as ruby's [`none?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_none.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [ ] when a class is passed as an argument returns true if none of the collection is a member of such class
```ruby
array = Array.new(100) { rand(0...9) }
array.my_none?(String) == true #true
```
- [ ] when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern
```ruby
words = %w[dog door rod blade]
words.my_none?(5) == words.none?(5) #true
```
- [ ] Regex is passed as an argument returns true only if none of the collection matches the Regex
```ruby
 words.my_none?(/z/) == words.none?(/z/) #true
 words.my_none?(/d/) == words.none?(/d/) #true
```

whiz25 commented on enumerables.rb:
```diff
@@ -0,0 +1,114 @@
+      i += 1
+    end
+    accumulator
+  end
```
- [ ] should be identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [ ] when a symbol is specified combines each element of the collection by applying the symbol as a named method
```ruby
array = Array.new(100) { rand(0...9) }
array.my_inject(:+) == array.inject(:+) #true
```

tirthajyoti-ghosh replied:
`array = Array.new(100) { rand(0...9) }`
`array.my_inject(:+) == array.inject(:+) #true`
The code above returns to true in my machine without any changes done after code review.

tirthajyoti-ghosh committed:
updated #my_all?

tirthajyoti-ghosh committed:
added support for Regexp, Class, Pattern in #my_all?, my_any?, #my_none?updated test file

tirthajyoti-ghosh committed:
reduced module length

tirthajyoti-ghosh commented:
Thanks, @whiz25 for the extensive code review.

Changes are done after code review:
- added Regexp, Class and Pattern support in #my_all?, #my_any?, #my_none?

*Changes requested on #my_inject method is already done and the test-case suggested returns to `true`, so no changes done there

MauricioRobayo reviewed (changes requested):
Very nice project!

It looks great and is nicely coming into place, but still need to fix some issues before continuing:

* [ ] `my_inject` is not behaving as `inject`, for example: `[1, 2, 3].my_inject(3, :*) == [1, 2, 3].inject(3, :*)`.
* [ ] `my_any?` should return false when no block or argument is given if the collection only contains false or nil.

Status: __Request changes__.

tirthajyoti-ghosh committed:
updated #my_inject, #my_any?, #my_each and test file

fixed #my_inject to prperly compute 2 params when no block given. fixed #my_any? to return false when no block or param is given if the collection only contains false or nil. added range support in #my_each.

tirthajyoti-ghosh commented:
Thanks, @MauricioRobayo for pointing out the above failing test cases.

These are the changes done after the 2nd code review:
- Fixed #my_inject to properly compute 2 params when no block is given.
- Fixed #my_any? to return false when no block or param is given if the collection only contains false or nil.
- Added range support in #my_each.

kobiyoyo reviewed (approved):
# Project Approved :thumbsup:
## Hello @tirthajyoti-ghosh  ,  You have done a great job with this project. :clap:
